### PR TITLE
upgrade VectorizationBase

### PR DESCRIPTION
**Pull request by GiggleLiu**
@chriselrod I am trying to upgrade VectorizationBase to v0.21. In order to see the place that it errors, I removed several lines about `stridedpointer` first. It magically not show any error.

Can you please help me check whether I am doing it correctly?

### CODE REVIEW

**Comment by GiggleLiu** [comment]
Wait... The tests take more than 3 hours and is still running, I have to kill it:
image

Why is it so slow? I didn't see the same issue on my local host.

**Comment by chriselrod** [comment]
> Why is it so slow? I didn't see the same issue on my local host.

Did tests pass locally?
If so, I'd let it keep going. Maybe reduce the number of tests.

CI takes many times longer when it's getting code coverage info.

File: src/gemm.jl [hunk 1]
```diff
@@ -51,19 +51,8 @@ end
...
-
-@inline function VectorizationBase.stridedpointer(
-    ptr::Ptr{T}, ::StaticInt{C}, ::StaticInt{B}, ::Val{R}, strd::X, offsets::O
-) where {T<:Tropical,C,B,R,N,X<:Tuple{Vararg{Integer,N}},O<:Tuple{Vararg{Integer,N}}}
-    VectorizationBase.StridedPointer{T,N,C,B,R,X,O}(ptr, strd, offsets)
-end
-
 @inline function notropical(ptr::VectorizationBase.StridedPointer{Tropical{T},N,C,B,R,X,O}) where {T,N,C,B,R,X,O}
-    VectorizationBase.StridedPointer{T,N,C,B,R,X,O}(Ptr{T}(ptr.p), ptr.strd, ptr.offsets)
+    VectorizationBase.StridedPointer{T,N,C,B,R,X,O}(Ptr{T}(ptr.p), ptr.si)
```
Comment (chriselrod) [inline src/gemm.jl]: You could also call
```julia
stridedpointer(Ptr{T}(ptr.p), ptr.si, StaticInt{B}())
```
but this is fine.

**Comment by GiggleLiu** [comment]
> > Why is it so slow? I didn't see the same issue on my local host.
> 
> Did tests pass locally?
> If so, I'd let it keep going. Maybe reduce the number of tests.
> 
> CI takes many times longer when it's getting code coverage info.

Yes. It is not about the number of tests, it is mainly about the first several tests. The compiling is very slow. The test on my local host takes less than 3min. I do not think taking 3 hour is normal.

**Comment by chriselrod** [comment]
> Yes. It is not about the number of tests, it is mainly about the first several tests. The compiling is very slow. The test on my local host takes less than 3min. I do not think taking 3 hour is normal.

Same basic story with Octavian's and LoopVectorization's tests.
I think the slow down you describe is a bit extreme, even compared to those libraries, but sadly it is not exactly unexpected.

File: src/gemm.jl (hunk 1 above)
Comment (GiggleLiu) [reply src/gemm.jl:55]: But I get:
```julia
mydot: Error During Test at /home/leo/.julia/dev/TropicalGEMM/test/gemm.jl:34
  Test threw exception
  Expression: mydot(a, b) ≈ transpose(a) * b
  MethodError: no method matching LayoutPointers.StridedPointer{Float64, 1, 1, 0, (1,), Tuple{StaticInt{8}}, Tuple{StaticInt{0}}}(::Ptr{Float64}, ::ArrayInterface.StrideIndex{1, (1,), 1, Tuple{StaticInt{8}}, Tuple{StaticInt{0}}}, ::StaticInt{0})
  Closest candidates are:
    LayoutPointers.StridedPointer{T, N, C, B, R, X, O}(::Any, ::Any) where {T, N, C, B, R, X, O} at /home/leo/.julia/packages/LayoutPointers/NXt9T/src/stridedpointers.jl:55
```

**Comment by GiggleLiu** [comment]
Just removed the coverage from CI. It works! thanks for your hint~

**Comment by chriselrod** [comment]
You could try something like Octavian does:
https://github.com/JuliaLinearAlgebra/Octavian.jl/blob/ead8fbb53d55fa726d135a7b83af099acec57093/.github/workflows/ci.yml#L14-L58
It has a "test" and "coverage" job.

Then it [checks if code coverage is enabled](https://github.com/JuliaLinearAlgebra/Octavian.jl/blob/ead8fbb53d55fa726d135a7b83af099acec57093/test/test_suite_preamble.jl#L9) before [running the slowest tests](https://github.com/JuliaLinearAlgebra/Octavian.jl/blob/ead8fbb53d55fa726d135a7b83af099acec57093/test/runtests.jl#L28-L30). Something like that may help.

It'd let you test a bunch of different sizes (if that's a concern), while still getting a code coverage (under)estimate from some smaller / faster subset of tests.

**Pull request merged**